sftp/sftp.go

```diff
@@ -0,0 +1,186 @@
+package sftp
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+)
+
+const (
+	// PasswordAuth indicates password-based authentication
+	PasswordAuth = "passwordAuth"
+	// KeyAuth indicates key-based authentication
+	KeyAuth = "keyAuth"
+)
+
+// SSHConfig represents the configuration for SSH connection
+type SSHConfig struct {
+	Host       string
+	Port       int
+	User       string
+	AuthMethod string
+	PrivateKey string
+	Password   string // Password for password-based authentication
+}
+
+// SSHClient interface abstracts the SSH client
+type SSHClient interface {
+	Dial(network, addr string, config *ssh.ClientConfig) (*ssh.Client, error)
+}
+
+// SFTPClient interface abstracts the SFTP client
+type SFTPClient interface {
+	CreateNew(client *ssh.Client) (*sftp.Client, error)
+	UploadFile(localFilePath, remoteDir string) error
+	DownloadFile(remoteFilePath, localDir string) error
+	DeleteFile(remoteFilePath string) error
+}
+
+// SSHClientImpl is a real implementation of SSHClient
+type SSHClientImpl struct{}
+
+// Dial establishes an SSH connection
+func (r *SSHClientImpl) Dial(network, addr string, config *ssh.ClientConfig) (*ssh.Client, error) {
+	return ssh.Dial(network, addr, config)
+}
+
+// SFTPClientImpl is a real implementation of SFTPClient
+type SFTPClientImpl struct {
+	client *sftp.Client
+}
+
+// Create creates a file on the remote server
+func (r *SFTPClientImpl) Create(remoteFilePath string) (io.WriteCloser, error) {
+	return r.client.Create(remoteFilePath)
+}
+
+// Create creates an SFTP client from an existing SSH client
+func (r *SFTPClientImpl) CreateNew(client *ssh.Client) (*sftp.Client, error) {
+	return sftp.NewClient(client)
+}
+
+// NewSSHClient establishes an SSH connection and returns an SSH client
+func NewSSHClient(config *SSHConfig, client SSHClient) (*ssh.Client, error) {
+	if config == nil {
```

Comment:
move config preparation to a separate function so that it can be tested and after that there won't be much logic left in this function so we are good

Reply:
So, we will not test the SSH Client creation?

Reply:
Addressed